### PR TITLE
Improve getItem

### LINKED ISSUE #49: Fix all the block.getItem() problems

### PR DESCRIPTION
**Pull request by cltnschlosser**
Fixes #49 

Should improve getItem, this mimics getPickBlock.
Adds a different message when there is no builder.
Some general improvements to code quality.

Review please @Minecolonies/maintainers

### CODE REVIEW

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 1]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
+            return Item.getItemFromBlock(Blocks.wooden_slab);
         }
+        else
+        {
+            return Item.getItemFromBlock(blockState.getBlock());
+        }
+    }
 
-        return item;
+    private static int getDamageValue(Block block, IBlockState blockState)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MAJOR Reduce the number of returns of this method 9, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)
MAJOR The Cyclomatic Complexity of this method "getDamageValue" is 19 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 2]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
+        else if (block instanceof BlockSilverfish)
+        {
+            return block.getMetaFromState(blockState);
+        }
+        else if (block instanceof BlockSlab)
+        {
+            return block.damageDropped(blockState) & 7;
+        }
+        else if (block instanceof BlockTallGrass)
+        {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MAJOR This branch's code block is the same as the block for the branch on line 297. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1871)

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 3]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
+        }
+        else if (block instanceof BlockDoublePlant
+                && blockState.getValue(BlockDoublePlant.HALF) == BlockDoublePlant.EnumBlockHalf.LOWER)
+        {
+            //If upper part we can't do much here
+            return blockState.getValue(BlockDoublePlant.VARIANT).getMeta();
+        }
+        else if (block instanceof BlockNewLeaf)
+        {
+            return block.getMetaFromState(blockState) & 3;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MINOR Assign this magic number 3 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 4]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
+        Block block = blockState.getBlock();
+        if (item instanceof ItemBlock && !block.isFlowerPot())
+        {
+            block = Block.getBlockFromItem(item);
+        }
+
+        return new ItemStack(item, 1, getDamageValue(block, blockState));
+    }
+
+    private static Item getItem(IBlockState blockState)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MAJOR Reduce the number of returns of this method 31, down to the maximum allowed 5. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1142)

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 5]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
-     * @return output item.
+     * Mimics pick block.
+     * @param blockState the block and state we are creating an ItemStack for.
+     * @return ItemStack fromt the BlockState.
      */
-    public static Item getItemFromBlock(Block block)
+    public static ItemStack getItemStackFromBlockState(IBlockState blockState)
     {
-        Item item = null;
+        Item item = getItem(blockState);
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MINOR Local variable 'item' could be declared final [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3ALocalVariableCouldBeFinal)

File: src/main/java/com/minecolonies/util/BlockUtils.java [hunk 6]
```diff
@@ -109,138 +113,201 @@ public static boolean isWater(IBlockState iBlockState)
...
+        {
+            return 0;
+        }
+        else if (block instanceof BlockSilverfish)
+        {
+            return block.getMetaFromState(blockState);
+        }
+        else if (block instanceof BlockSlab)
+        {
+            return block.damageDropped(blockState) & 7;
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/util/BlockUtils.java]: MINOR Assign this magic number 7 to a well-named constant, and use the constant instead. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS109)

File: src/main/java/com/minecolonies/entity/EntityCitizen.java [hunk 7]
```diff
@@ -1230,7 +1232,8 @@ private void sendChat(String msg)
         statusMessages.put(msg, ticksExisted);
 
         LanguageHandler.sendPlayersMessage(
-                ServerUtils.getPlayersFromUUID(worldObj, getColony().getPermissions().getMessagePlayers()),
+                colony.getMessageEntityPlayers(),
+                //TODO does this need to go through the LanguageHandler#format?
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/EntityCitizen.java]: INFO Complete the task associated to this TODO comment. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS1135)

File: src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java [hunk 8]
```diff
@@ -187,11 +190,17 @@ public void attemptToFulfill(Colony colony)
             }
         }
 
-        if (sendMessage && !hasSentMessageForThisWorkOrder)
+        if (!hasBuilder && !hasSentMessageForThisWorkOrder)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java]: MAJOR Avoid if (x != y) ..; else ..; [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

File: src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java [hunk 9]
```diff
@@ -152,6 +152,7 @@ public boolean isValid(Colony colony)
     public void attemptToFulfill(Colony colony)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java]: MAJOR The Cyclomatic Complexity of this method "attemptToFulfill" is 11 which is greater than 10 authorized. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AMethodCyclomaticComplexity)

File: src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 10]
```diff
@@ -564,26 +534,43 @@ private AIState decorationStep()
...
 
-        Item item = Item.getItemFromBlock(block);
-        worker.setCurrentItemOrArmor(0, item != null ? new ItemStack(item, 1) : null);
+        placeBlockAt(block, blockState, coords);
 
-        if (placeBlock(new BlockPos(x, y, z), block, metadata))
+        return findNextBlockNonSolid();
+    }
+
+    private void placeBlockAt(Block block, IBlockState blockState, BlockPos coords) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

File: src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 11]
```diff
@@ -688,64 +659,57 @@ private boolean placeBlock(BlockPos pos, Block block, IBlockState metadata)
...
-
-        if (stack.getItem() == null)
+        ItemStack stack = BlockUtils.getItemStackFromBlockState(blockState);
+        if (stack == null)
         {
-            stack = new ItemStack(BlockUtils.getItemFromBlock(block));
+            Log.logger.warn("Block causes NPE: " + blockState.getBlock());
         }
 
         int slot = worker.findFirstSlotInInventoryWith(stack.getItem());
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java]: BLOCKER NullPointerException might be thrown as 'stack' is nullable here [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AS2259)

File: src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java [hunk 12]
```diff
@@ -627,38 +614,22 @@ else if (entity instanceof EntityMinecart)
...
-            if (stack.getItem() == null)
-            {
-                stack = new ItemStack(block);
-            }
-
-            if (stack.getItem() == null)
-            {
-                stack = new ItemStack(BlockUtils.getItemFromBlock(block));
-            }
+        if (block == Blocks.air) {
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/entity/ai/citizen/builder/EntityAIStructureBuilder.java]: MINOR Move this left curly brace to the beginning of next line of code. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3ALeftCurlyBraceStartLineCheck)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 14 issues
* BLOCKER 1 blocker
* MAJOR 6 major
* MINOR 6 minor
* INFO 1 info

Watch the comments in this conversation to review them.

#### 1 extra issue

Note: The following issues were found on lines that were not modified in the pull request. Because these issues can't be reported as line comments, they are summarized here:

1. MINOR [WorkOrderBuild.java#L9](https://github.com/Minecolonies/minecolonies/blob/e51b9c41cdd16de06d9b55f9294f8e6719c5b805/src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java#L9): Remove this unused import 'com.minecolonies.util.ServerUtils'. [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=squid%3AUselessImportCheck)

File: src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java [hunk 13]
```diff
@@ -187,11 +189,27 @@ public void attemptToFulfill(Colony colony)
...
+    }
+
+    private void sendBuilderMessage(Colony colony, boolean hasBuilder, boolean sendMessage)
+    {
+        if (hasSentMessageForThisWorkOrder)
+        {
+            return;
+        }
+
+        if (!hasBuilder)
```
Comment (MinecoloniesSonar) [inline src/main/java/com/minecolonies/colony/workorders/WorkOrderBuild.java]: MAJOR Avoid if (x != y) ..; else ..; [rule](http://home.kk-sc.de:9000/coding_rules#rule_key=pmd%3AConfusingTernary)

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 7 issues
* MAJOR 4 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by MinecoloniesSonar** [comment]
SonarQube analysis reported 6 issues
* MAJOR 3 major
* MINOR 2 minor
* INFO 1 info

Watch the comments in this conversation to review them.

**Comment by Raycoms** [comment]
:+1: I'm okay with this.

**Comment by Kostronor** [comment]
:shipit:

**Pull request merged**